### ISSUE TITLE
As a developer, I need to hash pieces of the staged ledger reconstruction parts as soon as they arrive, to speed up it's verification

### ISSUE THREAD

**Issue by binier** [openmina/openmina#752]
This can be incredibly slow: https://github.com/openmina/openmina/blob/5810a039962c9e09c0e556939b427478a667fe26/node/src/transition_frontier/sync/ledger/staged/mod.rs#L51

Since we receive that struct, piece by piece in webnode, I want to start hashing them as soon as possible. This way we'll spread out computation, while we are waiting for the rest of the pieces from P2P.

**Comment by binier** [comment]
- [ ] Add validation logic for each received piece in p2p. This validation logic should calculate hash of the piece and cache it. These cached hashes must be used later as inputs are provided to the ledger service.
- [ ] **[Less Important]** In the initial response of staged ledger parts, hashes of all individual trees must be included, so that we can validate received data as soon as possible, to detect and disconnect malicious peer as soon as possible.

**Comment by binier** [comment]
After #734 pr, this issue is no longer relevant, since validation now is extremely fast, so no need parallelize it with fetching.
